### PR TITLE
RFQ accept price qualify bug

### PR DESCRIPTION
**Pull request by ffranr**
Fixes https://github.com/lightninglabs/taproot-assets/issues/1361  

This PR corrects a bug where we called the wrong price oracle endpoints when evaluating the price from a quote accept message sent in response to a quote request message.  

Additionally, we enhance the mock price oracle to allow method call hooking.  

See the commit messages for details on the original issue and the correct approach.

### CODE REVIEW

**Comment by ZZiigguurraatt** [comment]
One thing I realized when testing this fix is that when selling an asset, with `experimental.rfq.acceptpricedeviationppm=50000`, SellPrice>=(1-.05)*BuyPrice instead of BuyPrice<=(1+.05)*SellPrice . Wondering if we should make this clearer in the documentation?

**Comment by ZZiigguurraatt** [comment]
Can we get a more informative error message than

`peer 03bb1c09a26fbb1b5799b0263eed8e516fb5fbe141d93104ed19211a429f6065bd sent back an invalid quote, status: INVALID_ASSET_RATES`

That says that the rate was over the `experimental.rfq.acceptpricedeviationppm` limit? Like I can't tell if it is invalid because the quote isn't for a large enough amount of assets, the expiry time is past, etc..

**Comment by ZZiigguurraatt** [comment]
> One thing I realized when testing this fix is that when selling an asset with `SendPayment`, with `experimental.rfq.acceptpricedeviationppm=50000`, SellPrice>=(1-.05)*BuyPrice instead of BuyPrice<=(1+.05)*SellPrice . Wondering if we should make this clearer in the documentation?


Note, when buying an asset with `AddInvoice`, it works the same as above and SellPrice>=(1-.05)*BuyPrice. Given how SendPayment works, I would have thought it would be the opposite and be BuyPrice<=(1+.05)*SellPrice. I guess the seller is always the one that trigger's the error and the seller's error is just propagated back to the buyer when `AddInvoice` is run?

When using `AddInvoice` does the buyer give the edge node/seller a "price" or just a "price hint" when the RFQ initially starts? When using `AddInvoice` and the seller rejects, does the seller have any opportunity to make a counter offer? Just trying to understand so I know I'm testing properly.

File: rfq/oracle.go [hunk 1]
```diff
@@ -428,21 +428,6 @@ func NewMockPriceOracle(expiryDelay,
 	}
 }
 
-// NewMockPriceOracleSatPerAsset creates a new mock price oracle with a
-// specified satoshis per asset rate.
-//
-// TODO(ffranr): Remove in favour of NewMockPriceOracle.
-func NewMockPriceOracleSatPerAsset(expiryDelay uint64,
```
Comment (guggero) [inline rfq/oracle.go:435]: Hmm, so this was actually broken in 4e291e5e and we didn't realize back then. But the meaning between `rfqCfg.MockOracleAssetsPerBTC` and `rfqCfg.MockOracleSatsPerAsset` is different, so one of these functions is now incorrect, probably this one (since a rate is assets per BTC, so we'd need to convert things from sats per asset to that rate, which will be imprecise).
Or we completely remove the whole config flag for the sats per asset way of configuring the mock oracle.

File: rfq/negotiator_test.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+	negotiator, err := NewNegotiator(NegotiatorCfg{
+		PriceOracle:             mockPriceOracle,
+		OutgoingMessages:        make(chan rfqmsg.OutgoingMsg, 1),
+		AcceptPriceDeviationPpm: tc.AcceptPriceDeviationPpm,
+		ErrChan:                 errChan,
+	})
+	require.NoError(t, err)
+
+	// Define the finalise callback function.
+	finalise := func(
```
Comment (guggero) [inline rfq/negotiator_test.go:87]: nit: format like other function definitions instead? Her and in other test below.

```
	finalise := func(msg rfqmsg.BuyAccept,
		event fn.Option[InvalidQuoteRespEvent]) {
```

File: rfq/negotiator_test.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	"github.com/lightningnetwork/lnd/routing/route"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+// testCaseIncomingSellAccept is a test case for the handling of an incoming
+// sell accept message.
+type testCaseIncomingSellAccept struct {
+	// Name is the name of the test case.
+	Name string
```
Comment (guggero) [inline rfq/negotiator_test.go:21]: nit: do all of these fields need to be exported? Since the main struct isn't?

File: rfq/negotiator_test.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+// TestHandleIncomingBuyAccept tests the handling of an incoming buy accept
+// message.
+func TestHandleIncomingBuyAccept(t *testing.T) {
+	defaultQuoteExpiry := time.Now().Add(time.Hour)
+
+	testCases := []testCaseIncomingBuyAccept{
+		{
+			Name: "accept price within bounds",
```
Comment (guggero) [inline rfq/negotiator_test.go:302]: Should we also add a test case where the price is within the bounds, but just barely? Instead of only being exactly the same? Here and in the test above.

**Review by guggero — COMMENTED** [review]
Changes to the RFQ code look good to me. Nice investigation and fixes!

I noticed we broke something with the mock oracle in an earlier PR and I think we should fix that again or remove things completely.

File: rfq/negotiator_test.go (hunk 3 above)
Comment (ffranr) [reply rfq/negotiator_test.go:21]: I'll make them all non-exported.

File: rfq/oracle.go (hunk 1 above)
Comment (ffranr) [reply rfq/oracle.go:435]: I'll leave the old constructor in and I'll fix the bug. We need to convert sats per asset to asset per btc correctly.

Comment (ffranr) [reply rfq/oracle.go:435]: I've added a TODO in this PR to fix this oversight.

File: rfq/negotiator_test.go (hunk 4 above)
Comment (ffranr) [reply rfq/negotiator_test.go]: Done.

But see this PR for an explanation of the numbers: https://github.com/lightninglabs/taproot-assets/pull/1377

File: rfq/negotiator_test.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/rfqmath"
+	"github.com/lightninglabs/taproot-assets/rfqmsg"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+// testCaseIncomingSellAccept is a test case for the handling of an incoming
```
[the remark this answers is not in the document]
Comment (ffranr) [reply rfq/negotiator_test.go:17]: Good idea. I might look into that later. We have rapid tests for the tolerance method already. They might be handy for the expiry.

**Comment by ffranr** [comment]
@ZZiigguurraatt Thank you for your comments, I've gone through them. I think we will need to consider them separately from this bug. I'll think about what you've said some more tomorrow.

File: rfq/negotiator_test.go [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+	defaultQuoteExpiry := time.Now().Add(time.Hour)
+
+	testCases := []testCaseIncomingSellAccept{
+		{
+			name: "accept price just within bounds 1",
+			incomingSellAcceptRate: rfqmsg.AssetRate{
+				Rate:   rfqmath.NewBigIntFixedPoint(1000, 0),
+				Expiry: defaultQuoteExpiry,
+			},
+			priceOracleAskPrice: rfqmsg.AssetRate{
```
Comment (guggero) [inline rfq/negotiator_test.go:140]: Nice, seems like we caught some edge cases in the tolerance calculation with these test cases?

**Review by guggero — APPROVED** [review]

File: rfq/negotiator_test.go (hunk 6 above)
Comment (ffranr) [reply rfq/negotiator_test.go:140]: From my POV, it forced us to be clear about the methodology we use to check tolerance.
